pkg/validation/options.go

```diff
@@ -219,69 +196,69 @@ func Validate(o *options.Options) error {
 
 func parseProviderInfo(o *options.Options, msgs []string) []string {
 	p := &providers.ProviderData{
-		Scope:            o.Scope,
-		ClientID:         o.ClientID,
-		ClientSecret:     o.ClientSecret,
-		ClientSecretFile: o.ClientSecretFile,
-		Prompt:           o.Prompt,
-		ApprovalPrompt:   o.ApprovalPrompt,
-		AcrValues:        o.AcrValues,
+		Scope:            o.Providers[0].Scope,
+		ClientID:         o.Providers[0].ClientID,
+		ClientSecret:     o.Providers[0].ClientSecret,
+		ClientSecretFile: o.Providers[0].ClientSecretFile,
+		Prompt:           o.Providers[0].Prompt,
+		ApprovalPrompt:   o.Providers[0].ApprovalPrompt,
+		AcrValues:        o.Providers[0].AcrValues,
 	}
-	p.LoginURL, msgs = parseURL(o.LoginURL, "login", msgs)
-	p.RedeemURL, msgs = parseURL(o.RedeemURL, "redeem", msgs)
-	p.ProfileURL, msgs = parseURL(o.ProfileURL, "profile", msgs)
-	p.ValidateURL, msgs = parseURL(o.ValidateURL, "validate", msgs)
-	p.ProtectedResource, msgs = parseURL(o.ProtectedResource, "resource", msgs)
+	p.LoginURL, msgs = parseURL(o.Providers[0].LoginURL, "login", msgs)
+	p.RedeemURL, msgs = parseURL(o.Providers[0].RedeemURL, "redeem", msgs)
+	p.ProfileURL, msgs = parseURL(o.Providers[0].ProfileURL, "profile", msgs)
+	p.ValidateURL, msgs = parseURL(o.Providers[0].ValidateURL, "validate", msgs)
+	p.ProtectedResource, msgs = parseURL(o.Providers[0].ProtectedResource, "resource", msgs)
 
 	// Make the OIDC Verifier accessible to all providers that can support it
 	p.Verifier = o.GetOIDCVerifier()
 
-	p.SetAllowedGroups(o.AllowedGroups)
+	p.SetAllowedGroups(o.Providers[0].AllowedGroups)
 
-	provider := providers.New(o.ProviderType, p)
+	provider := providers.New(o.Providers[0].ProviderType, p)
 	if provider == nil {
-		msgs = append(msgs, fmt.Sprintf("invalid setting: provider '%s' is not available", o.ProviderType))
+		msgs = append(msgs, fmt.Sprintf("invalid setting: provider '%s' is not available", o.Providers[0].ProviderType))
 		return msgs
 	}
 	o.SetProvider(provider)
 
 	switch p := o.GetProvider().(type) {
```

Comment:
Not sure if it belongs here in this PR or is too much scope creep and belongs in another PR:

I hate this switch statement 😄 

I think all of this should be a `Configure(opts *options.Options)` method that the `Provider` interface mandates. Then each provider can implement the custom logic they need in their code instead of way out here separate from the rest of their business logic.

The only challenge is the few that manipulate the validate `msg` in the error case.

Reply:
Actually... looks like you might have that concern handled in your new validator code 🤔

Reply:
@JoelSpeed Any ideas on this? If we moved configuration setup into `providers` - where should mandatory configuration validation business logic reside? In the provider as well? Or in the `validation` package still but organized so it is clear where provider specific validations live?

Reply:
I would expect Configure to return errors and if there's any specific validation we want for the options, that can be done in `validation` as well. We did this for a bunch of the other refactors recently. Yes the logic is duplicated, but it makes the two parts completely disjoint. Validation is best effort after all

Reply:
If you guys don't mind, let's move this discussion/implementation to the next PR in this series? Just so that we have logical separation between 'multiple providers - first step including validation' and 'multiple providers - provider specific validation logic'. Please un-resolve if you disagree..

Reply:
Sounds good to me!